app/src/main/java/com/example/a17255/a8project/GamePlay.java

```diff
@@ -92,26 +160,173 @@ public MyCanvas(Context context) {
+//                    change direction when collide, u should try to keep away from the dark gary obstcle if u want to achieve high score
+                    currentSpeedX = Math.abs(currentSpeedX) - 1000;
+                    currentSpeedY = Math.abs(currentSpeedY) - 1000;
+                    obstaclePaint3.setColor(Color.CYAN);
+                }
+                if (Math.pow(mCurrentX - obstacleX4, 2) + Math.pow(mCurrentY - obstacleY4, 2) <=
+                        Math.pow(BALL_RADIUS + OBSTACLE_RADIUS_2, 2)) {
+//                    reduce score when collide
+                    score -= 7;
+
+                    currentSpeedX = Math.abs(currentSpeedX) - 1000;
+                    currentSpeedY = Math.abs(currentSpeedY) - 1000;
+                    obstaclePaint4.setColor(Color.CYAN);
+                }
+                if (Math.pow(mCurrentX - obstacleX5, 2) + Math.pow(mCurrentY - obstacleY5, 2) <=
+                        Math.pow(BALL_RADIUS + OBSTACLE_RADIUS_2, 2)) {
+//                    reduce score when collide
+                    score -= 7;
+
+                    currentSpeedX = Math.abs(currentSpeedX) - 1000;
+                    currentSpeedY = Math.abs(currentSpeedY) - 1000;
+                    obstaclePaint5.setColor(Color.CYAN);
+                }
+                if (Math.pow(mCurrentX - obstacleX3, 2) + Math.pow(mCurrentY - obstacleY3, 2) >
+                        Math.pow(BALL_RADIUS + OBSTACLE_RADIUS_2, 2)) {
+                    obstaclePaint3.setColor(Color.DKGRAY);
+                }
+                if (Math.pow(mCurrentX - obstacleX4, 2) + Math.pow(mCurrentY - obstacleY4, 2) >
+                        Math.pow(BALL_RADIUS + OBSTACLE_RADIUS_2, 2)) {
+                    obstaclePaint4.setColor(Color.DKGRAY);
+                }
+                if (Math.pow(mCurrentX - obstacleX5, 2) + Math.pow(mCurrentY - obstacleY5, 2) >
+                        Math.pow(BALL_RADIUS + OBSTACLE_RADIUS_2, 2)) {
+                    obstaclePaint5.setColor(Color.DKGRAY);
+                }
+
+
+//                detect the collision between the ball and the target
+                if (Math.pow(mCurrentX - targetX, 2) + Math.pow(mCurrentY - targetY, 2) <=
+                        Math.pow(BALL_RADIUS + TARGET_RADIUS, 2)) {
+
+//                    change score when collide
+                    currentSpeedX = 0;
+                    currentSpeedY = 0;
+                    score += 200;
+
+//                    save score in the SharedPreferences
+                    saveScoreInSharedPreferences(score);
+
+//                    alert dialog
+                    confirmFireMissiles();
+
+
+
+                }
+
+
+
+
+
+
+//                display score
+                drawScore(canvas);
+
+
+//                draw ball after fling
+
+                mCurrentX = mCurrentX + currentSpeedX / 4000 * DURATION_PER_FRAME_IN_MS;
+                mCurrentY = mCurrentY + currentSpeedY / 4000 * DURATION_PER_FRAME_IN_MS;
```

Comment:
why 4000?
currentSpeedY is pixel per second
currentSpeedY/1000 = pixel per mili second
currentSpeedY / 1000 * DURATION_PER_FRAME_IN_MS = # of pixel moved per frame

Reply:
It is way to fast when I try to play it... The ball can easily run all over the screen with only shadow haha. Actually I should do 1000 * 4 and name the 4 as XXXsetSpeedRate to make it more readable.